Mathlib/LinearAlgebra/FiniteDimensional.lean

```diff
@@ -0,0 +1,1605 @@
+    -- express the summand using `shift`.
+    change (∑ x : V in t.erase x₀, (fun e => g e • e) (shift x)) = 0
+    -- porting note: last argument can't be inferred
+    rw [← sum_map _ shift (fun e => g e • e)]
+    exact gsum
+  · show (∑ e : V in t, f e) = 0
+    -- Again we split off the `x₀` term,
+    -- observing that it exactly cancels the other terms.
+    rw [← insert_erase m, sum_insert (not_mem_erase x₀ t)]
+    dsimp
+    rw [if_pos rfl]
+    conv_lhs =>
+      congr
+      rfl
+      apply_congr
+      rfl
+      rw [if_neg (show _ ≠ x₀ from (mem_erase.mp ‹_›).1)]
+    exact neg_add_self _
+  · show ∃ (x : V), x ∈ t ∧ f x ≠ 0
+    -- We can use x₁ + x₀.
+    refine' ⟨x₁ + x₀, _, _⟩
+    · rw [Finset.mem_map] at x₁_mem
+      rcases x₁_mem with ⟨x₁, x₁_mem, rfl⟩
+      rw [mem_erase] at x₁_mem
+      simp only [x₁_mem, sub_add_cancel, Function.Embedding.coeFn_mk]
+    · dsimp only
+      rwa [if_neg, add_sub_cancel]
+      rw [add_left_eq_self]
+      rintro rfl
+      simp only [sub_eq_zero, exists_prop, Finset.mem_map, Embedding.coeFn_mk, eq_self_iff_true,
+        mem_erase, not_true, exists_eq_right, Ne.def, false_and_iff] at x₁_mem
+#align finite_dimensional.exists_nontrivial_relation_sum_zero_of_rank_succ_lt_card FiniteDimensional.exists_nontrivial_relation_sum_zero_of_rank_succ_lt_card
+
+section
+
+variable {L : Type _} [LinearOrderedField L]
+
+variable {W : Type v} [AddCommGroup W] [Module L W]
+
+/-- A slight strengthening of `exists_nontrivial_relation_sum_zero_of_rank_succ_lt_card`
+available when working over an ordered field:
+we can ensure a positive coefficient, not just a nonzero coefficient.
+-/
+theorem exists_relation_sum_zero_pos_coefficient_of_rank_succ_lt_card [FiniteDimensional L W]
+    {t : Finset W} (h : finrank L W + 1 < t.card) :
+    ∃ f : W → L, (∑ e in t, f e • e) = 0 ∧ (∑ e in t, f e) = 0 ∧ ∃ x ∈ t, 0 < f x := by
+  obtain ⟨f, sum, total, nonzero⟩ := exists_nontrivial_relation_sum_zero_of_rank_succ_lt_card h
+  exact ⟨f, sum, total, exists_pos_of_sum_zero_of_exists_nonzero f total nonzero⟩
+#align finite_dimensional.exists_relation_sum_zero_pos_coefficient_of_rank_succ_lt_card FiniteDimensional.exists_relation_sum_zero_pos_coefficient_of_rank_succ_lt_card
+
+end
+
+end
+
+/-- In a vector space with dimension 1, each set {v} is a basis for `v ≠ 0`. -/
+@[simps]
+noncomputable def basisSingleton (ι : Type _) [Unique ι] (h : finrank K V = 1) (v : V)
+    (hv : v ≠ 0) : Basis ι K V :=
+  let b := basisUnique ι h
+  let h : b.repr v default ≠ 0 := mt basisUnique.repr_eq_zero_iff.mp hv
+  Basis.ofRepr
+    { toFun := fun w => Finsupp.single default (b.repr w default / b.repr v default)
+      invFun := fun f => f default • v
+      map_add' := by simp [add_div]
+      map_smul' := by simp [mul_div]
+      left_inv := fun w => by
+        apply_fun b.repr using b.repr.toEquiv.injective
+        -- porting note: was `apply_fun Equiv.finsuppUnique`
+        refine Equiv.finsuppUnique.injective ?_
```

Comment:
@kmill, any chance you want to look at this as our resident `apply_fun` expert? :-)

Reply:
I think there's a `TODO` comment in `apply_fun` about this

Reply:
Couldn't find one?

Reply:
The "`apply_fun` could not apply `Equiv.finsuppUnique` to the main goal" error is unrelated to the TODO that used to be there about `Equiv` (before the most recent `apply_fun` PR).

The issue is that `apply_fun` is missing some sort of expression normalization for the target before determining what the goal relation is.
```suggestion
        -- porting note: need `dsimp only` to get `apply_fun` to see that the goal is an equality (this is a bug in `apply_fun`).
        dsimp only
        apply_fun Equiv.finsuppUnique
```
I'll try to fix this.

Reply:
#3621